### PR TITLE
Initial attempt at ReplaceReturnTypeFix.

### PR DESCRIPTION
**Pull request by nojaf**
Hello @auduchinok, this is a first attempt to add a quick fix action for the `Type constraint mismatch` error when a defined return type differs from the actual type of an expression.

ReplaceReturnType

Please let me know what the best next steps could be here.
I imagine unit tests and edge cases.

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+type ReplaceReturnTypeFix(expr: IFSharpExpression, fcsErrorMessage: string, replaceFirstType: bool) =
+    inherit FSharpQuickFixBase()
+    
+    let rec visitParent (expr: IFSharpExpression): IFSharpExpression option =
+        match expr.Parent with
+        | :? ISequentialExpr as seqExpr ->
+            if seqExpr.Expressions.Last() <> expr then None else
+            visitParent seqExpr
+        | :? IMatchClause as clause ->
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs:19]: Getting the match clause parent like this is not good here. Is it the `when` expression or is it the expression after `->`? We're only interested in the last one here.
There're generated methods that help with such navigation, take a look at `MatchClauseNavigator` and similar classes.

**Review by auduchinok — COMMENTED** [review]
@nojaf Could you try to unify the tree traversal with the logic in `RecordCtorReference.GetFcsSymbol`?
https://github.com/JetBrains/resharper-fsharp/blob/net222/ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs would be a good place to put it to.

**Comment by nojaf** [comment]
@auduchinok, I'm slowly getting the hang of this.
Let me know where you still see gaps.
I'm sure there are some more cases regarding `ITypeUsage`.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 2]
```diff
@@ -35,5 +35,106 @@ public static class FSharpExpressionUtil
...
+          case ILiteralExpr literalExpr:
+            currentExpr = literalExpr;
+            parent = literalExpr.Parent;
+            break;
+          case IMatchClause matchClause:
+            if (matchClause.Expression == currentExpr)
+            {
+              currentExpr = matchClause;
+              parent = matchClause.Parent;
+            }
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:59]: This looks a bit like reimplementation of `MatchClauseNavigator.GetByExpression` and similar methods. For going up the tree they are preferred way in many cases, since they allow skipping intermediate nodes (when the rule allows) and make it easier to change the tree structure and grammar.

With multi-step navigators used, so it could be done in a way like this:
```csharp
while (true)
{
  if (MatchExprNavigator.GetByClauseExpression(currentExpr) is { } matchExpr)
  {
    currentExpr = matchExpr;
    continue;
  }

  if (SequentialExprNavigator.GetByExpression(currentExpr) is { } seqExpr &&
      seqExpr.ExpressionsEnumerable.Last() == currentExpr)
  {
    currentExpr = seqExpr;
    continue;
  }

  // other navigators go 

  return currentExpr;
}
```

This way you don't have to deal with non-expr parent nodes (like the chameleon expr case above). This is also arguably clearer than a big switch statement. 🙂

Getting parent multiple times is not a problem performance-wise here, as it's a simple field lookup and we only run this code when a caret is at a matching error highlighting type.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 3]
```diff
@@ -35,5 +35,106 @@ public static class FSharpExpressionUtil
...
+              currentExpr = null;
+              parent = null;
+            }
+            break;
+          case IIfThenElseExpr ifThenElseExpr:
+            if (ifThenElseExpr.ThenExpr == currentExpr || ifThenElseExpr.ElseExpr == currentExpr)
+            {
+              currentExpr = ifThenElseExpr;
+              parent = ifThenElseExpr.Parent;
+            }
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:83]: Please note that the overall type of an `if` expression depends on a presence of an `else` branch:
```fsharp
if true then 1 // the overall type is `unit`
if true then 1 else 2 // the overall type is `int`
```

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 4]
```diff
@@ -35,5 +35,106 @@ public static class FSharpExpressionUtil
...
+            else
+            {
+              currentExpr = null;
+              parent = null;
+            }
+            break;
+          case IForExpr forExpr:
+            if (forExpr.DoExpression == currentExpr)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:91]: Likewise, `for` and `while` expressions are always considered `unit`, so the `do` expression doesn't influence the overall type and we can't go up here.

File: ReSharper.FSharp/test/data/features/quickFixes/replaceReturnType/Replace type with attribute.fs [hunk 5]
```diff
@@ -0,0 +1,4 @@
+type FooAttribute() =
+    inherit System.Attribute()
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/quickFixes/replaceReturnType/Replace type with attribute.fs:2]: Minor: maybe it could use an existing attribute instead of defining one to keep the test simpler.

File: ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+           && isNotNull binding.ReturnTypeInfo.ReturnType then
+            let factory = binding.CreateElementFactory()
+            let typeUsage = factory.CreateTypeUsage(replacementTypeName)
+                
+            match binding.ReturnTypeInfo.ReturnType with
+            | :? INamedTypeUsage as ntu ->
+                ModificationUtil.ReplaceChild(ntu, typeUsage) |> ignore
+            | :? IFunctionTypeUsage as ftu ->
+                ftu.SetReturnTypeUsage(typeUsage) |> ignore
+            | _ -> ()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs:75]: This doesn't cover parens:
```fsharp
let a: (int) = ""
let b: (int * int) = "", 1
```

**Review by auduchinok — COMMENTED** [review]
@nojaf The overall approach is much better now, I left a few comments here and there.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 7]
```diff
@@ -35,5 +35,106 @@ public static class FSharpExpressionUtil
...
+            else
+            {
+              currentExpr = null;
+              parent = null;
+            }
+            break;
+          case IFSharpExpression fSharpExpression:
+            currentExpr = fSharpExpression;
+            parent = fSharpExpression.Parent;
+            break;
```
[the remark this answers is not in the document]
Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:129]: It actually will, as I'm ignoring the `IRecordFieldBinding` parent in the default case.
I'll add a negative test for it.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 8]
```diff
@@ -35,5 +35,106 @@ public static class FSharpExpressionUtil
...
+      while (parent != null)
+      {
+        switch (parent)
+        {
+          case IChameleonExpression:
+            return currentExpr as IFSharpExpression;
+          case ILiteralExpr literalExpr:
+            currentExpr = literalExpr;
+            parent = literalExpr.Parent;
+            break;
```
[the remark this answers is not in the document]
Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:53]: Yes, this doesn't make sense indeed.
I'll remove it.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:129]: Yes, right, indeed. Probably no much need for the test, there are many similar cases throughout the grammar, but it's up to you.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 9]
```diff
@@ -35,5 +36,54 @@ public static class FSharpExpressionUtil
...
+        }
+        
+        if (SequentialExprNavigator.GetByExpression(currentExpr) is { } seqExpr &&
+            seqExpr.ExpressionsEnumerable.Last() == currentExpr)
+        {
+          currentExpr = seqExpr;
+          continue;
+        }
+
+        if (IfExprNavigator.GetByConditionExpr(currentExpr) is { } ifExpr &&
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:59]:
```fsharp
let _: int = if true then 1 else "" // the error is in the else branch
let _: int = if 1 then 1 else 2 // we should't go up here
```

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 10]
```diff
@@ -35,5 +36,54 @@ public static class FSharpExpressionUtil
...
+            ifExpr.ElseExpr != null)
+        {
+          currentExpr = ifExpr;
+          continue;
+        }
+
+        if (BinaryAppExprNavigator.GetByArgument(currentExpr) is { } binaryAppExpr)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:66]: We likely need to be more cautious here. Consider this:
```fsharp
let _: int = 1 + ""
```

I'd also check if the logic is the same for piping. For some reason I think there might be some quirks with binary operators and type conversions.

In general, it's usually better to not provide quick fix in cases were we aren't sure we work correctly than breaking the user code in an unexpected way. It's better to gradually enable cases later than to ship it as soon as possible and to deal with a situation where you can't easily fix all the corner cases and don't want to disable the feature too.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 11]
```diff
@@ -35,5 +36,54 @@ public static class FSharpExpressionUtil
...
+        }
+
+        if ((TryWithExprNavigator.GetByTryExpression(currentExpr) ??
+             TryWithExprNavigator.GetByClauseExpression(currentExpr)) is { } tryWithExpr)
+        {
+          currentExpr = tryWithExpr;
+          continue;
+        }
+
+        return currentExpr;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:85]: The previous implementation also had a `let` expressions part.
```fsharp
let x: int =
    let _ = 1
    ""
```

File: ReSharper.FSharp/test/src/FSharp.Intentions.Tests/src/QuickFixes/ReplaceReturnTypeTest.fs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+    override x.RelativeTestDataPath = "features/quickFixes/replaceReturnType"
+
+    [<Test>] member x.``Replace type in Constraint Mismatch``() = x.DoNamedTest()
+    [<Test>] member x.``Replace type in infix application``() = x.DoNamedTest()
+    [<Test>] member x.``Replace type in Sequential``() = x.DoNamedTest()
+    [<Test>] member x.``Replace type in MatchClause``() = x.DoNamedTest()
+    [<Test>] member x.``Replace type with attribute``() = x.DoNamedTest()
+    [<Test>] member x.``Replace type in TryWith``() = x.DoNamedTest()
+    [<Test>] member x.``Replace type in MatchLambda``() = x.DoNamedTest()
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Intentions.Tests/src/QuickFixes/ReplaceReturnTypeTest.fs:19]: I'd probably drop `Replace type` in all tests, as this is covered by the class name.

I also tend to use a simple name with a number at the end, as often another related test case is added later and it's easier to group them by using a similar name.

File: ReSharper.FSharp/test/src/FSharp.Intentions.Tests/src/QuickFixes/ReplaceReturnTypeTest.fs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    [<Test; NoHighlightingFound>] member x.``Replace type - Negative 01``() = x.DoNamedTest()
+    [<Test; NotAvailable>] member x.``Replace type - Negative 02``() = x.DoNamedTest()
+    [<Test; NotAvailable>] member x.``Replace type - Negative 03``() = x.DoNamedTest()
+    [<Test; NotAvailable>] member x.``Replace type - Negative 04``() = x.DoNamedTest()
+    [<Test; NoHighlightingFound>] member x.``Replace type - Negative 05``() = x.DoNamedTest()
+    [<Test; NotAvailable>] member x.``Replace type - Negative 06``() = x.DoNamedTest()
+    [<Test; NotAvailable>] member x.``Replace type - Negative 07``() = x.DoNamedTest()
+    [<Test; NoHighlightingFound>] member x.``Replace type - Negative 08``() = x.DoNamedTest()
+    [<Test; NotAvailable>] member x.``Replace type - Negative 09``() = x.DoNamedTest()
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Intentions.Tests/src/QuickFixes/ReplaceReturnTypeTest.fs:28]: It may be a bit clearer to group these tests by the expected error (and to add related cases easier later).
```fsharp
[<Test; NoHighlightingFound>] member x.``No highlighting 01``() = x.DoNamedTest()
[<Test; NoHighlightingFound>] member x.``No highlighting 02``() = x.DoNamedTest()
...
[<Test; NotAvailable>] member x.``Not available - Binary app 01``() = x.DoNamedTest()
[<Test; NotAvailable>] member x.``Not available - Typed pattern 01``() = x.DoNamedTest()
...
```

**Review by auduchinok — COMMENTED** [review]
Here goes another round. 🙂

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 14]
```diff
@@ -35,5 +36,58 @@ public static class FSharpExpressionUtil
...
+        {
+          currentExpr = lambdaExpr;
+          continue;
+        }
+
+        if (TupleExprNavigator.GetByExpression(currentExpr) is { } tupleExpr)
+        {
+          currentExpr = tupleExpr;
+          continue;
+        }
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:81]: Could you show an example where we need to go up from a tuple expression?

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 15]
```diff
@@ -35,5 +36,58 @@ public static class FSharpExpressionUtil
...
+        {
+          currentExpr = letOrUseExpr;
+          continue;
+        }
+
+        if (LambdaExprNavigator.GetByExpression(currentExpr) is { } lambdaExpr)
+        {
+          currentExpr = lambdaExpr;
+          continue;
+        }
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:75]: Likewise, can we go up from any lambda body?

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 16]
```diff
@@ -35,5 +36,58 @@ public static class FSharpExpressionUtil
...
+
+        if (TupleExprNavigator.GetByExpression(currentExpr) is { } tupleExpr)
+        {
+          currentExpr = tupleExpr;
+          continue;
+        }
+
+        if (BinaryAppExprNavigator.GetByArgument(currentExpr) is { } binaryAppExpr)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:83]:
```fsharp
let x: int = 1 + ""
```

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 14 above)
Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:81]: See test `Tuple return type.fs`

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 15 above)
Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:75]: See test `FunctionType 01.fs`.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 16 above)
Comment (nojaf) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:83]: Yes, still on my to-do list.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 15 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:75]: I see, thanks. We should carefully consider possible syntax and FCS types here:
```fsharp
let f: int -> double -> string = Unchecked.defaultof<_>
let g: int -> int -> string = f
```
```fsharp
let h: int * (int -> int) = 1, fun x -> ""
```

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 14 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:81]: Another interesting case:
```fsharp
let x: int -> int = 1, 1
```

File: ReSharper.FSharp/src/FSharp.Psi.Services/src/Daemon/Highlightings/FcsErrors.xml [hunk 17]
```diff
@@ -525,5 +525,18 @@
     <Behavour overlapResolvePolicy="NONE"/>
     <QuickFix>ReplaceReturnTypeFix</QuickFix>
   </Error>
-  
+
+  <Error staticGroup="FSharpErrors" name="TypeMisMatchTuplesHaveDifferingLengths" ID="FS0001: Type mismatch. Expecting">
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Services/src/Daemon/Highlightings/FcsErrors.xml:529]: Maybe we probably should check this inside the quick fix's `IsAvailable` method instead of creating different error types. I'll need to think about it.

Parsing the error also seem a too ad-hoc and a bit dangerous, it would be much better to work with FCS's `FSharpType` instead. It's not always possible to get that view of the expected type, though.

**Comment by nojaf** [comment]
Would it be an option to put some more logic into the `override this.IsAvailable` member to skip the more complicated scenarios?

File: ReSharper.FSharp/src/FSharp.Psi.Daemon/src/Stages/FcsErrorsStageProcessBase.fs [hunk 18]
```diff
@@ -184,6 +187,17 @@ type FcsErrorsStageProcessBase(fsFile, daemonProcess) =
...
+                let expr = nodeSelectionProvider.GetExpressionInRange(fsFile, range, false, null)
+                TypeDoesNotMatchTypeError(expectedType, actualType, expr, error.Message)
+
+            | Regex ifBranchSatisfyContextTypeRequirements [expectedType; actualType] ->
+                let expr = nodeSelectionProvider.GetExpressionInRange(fsFile, range, false, null)
+                IfExpressionNeedsTypeToSatisfyTypeRequirementsError(expectedType, actualType, expr, error.Message)
+
+            | Regex typeMisMatchTupleLengths [expectedType; actualType] ->
+                let expr = nodeSelectionProvider.GetExpressionInRange(fsFile, range, false, null)
+                TypeMisMatchTuplesHaveDifferingLengthsError(expectedType, actualType, expr, error.Message)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Daemon/src/Stages/FcsErrorsStageProcessBase.fs:200]: There's something wrong with reported types when item count mismatch:

```fsharp
let x: int * int * int = 1, ""
```

Screenshot 2022-05-24 at 17 23 33

Let's disable this case until the error is fixed?

File: ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        let refPat = binding.HeadPattern.As<IReferencePat>()
+        if isNull refPat then () else
+
+        let symbolUse = refPat.GetFcsSymbolUse()
+        if isNull symbolUse then () else
+
+        if isNotNull binding.ReturnTypeInfo
+           && isNotNull binding.ReturnTypeInfo.ReturnType then
+            let factory = binding.CreateElementFactory()
+            let typeUsage = factory.CreateTypeUsage(replacementTypeName)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs:66]: I consider this creation of a type usage from a text parsed in an error text to be a bit dangerous but we can try to use it for the time being.

File: ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+                | returnType -> returnType
+
+            match returnType with
+            | :? INamedTypeUsage as ntu ->
+                ModificationUtil.ReplaceChild(ntu, typeUsage) |> ignore
+            | :? IFunctionTypeUsage as ftu ->
+                match mostOuterParentExpr with
+                | :? ITupleExpr ->
+                    ModificationUtil.ReplaceChild(ftu, typeUsage) |> ignore
+                | _ -> ftu.SetReturnTypeUsage(typeUsage) |> ignore
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Intentions/src/QuickFixes/ReplaceReturnTypeFix.fs:79]: This doesn't handle this case correctly:

```fsharp
let x = 1, 2
let f: int -> int = x 
```

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs [hunk 21]
```diff
@@ -35,5 +36,58 @@ public static class FSharpExpressionUtil
 
     public static readonly Func<IFSharpExpression, bool> IsSimpleValueExpressionFunc = IsSimpleValueExpression;
     public static readonly Func<IFSharpExpression, bool> IsLiteralExpressionFunc = IsLiteralExpression;
+
+    // TODO: change name
+    public static IFSharpExpression GetOuterMostParentExpression(this IFSharpExpression expression)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:41]: Maybe we use something like `GetOutermostParentExpressionFromItsReturn` for now (though, I find it too long for to be good).

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 15 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:75]: Both these cases don't work correctly with the current implementation. We should disable going up from lambda expressions for now.

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 14 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:81]: We shouldn't go up here, but it's OK to go up in the quick fix (as you already seem to do).

File: ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs (hunk 16 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs:83]: Please remove this `if` for now 🙂

**Comment by nojaf** [comment]
My pleasure, thank you for all the guidance along the way!
